crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+            Operation::trunc => BigUint::min(
+                lhs_max,
+                BigUint::from(2_u32).pow(rhs_max.try_into().unwrap()) - BigUint::from(1_u32),
+            ),
+            //'a = b': a and b must be of same type.
+            Operation::ass => rhs_max,
+            Operation::nop | Operation::jne | Operation::jeq | Operation::jmp => todo!(),
+            Operation::phi => BigUint::max(lhs_max, rhs_max), //TODO operands are in phi_arguments, not lhs/rhs!!
+            Operation::eq_gate => BigUint::min(lhs_max, rhs_max),
+        }
+    }
+
+    // pub fn get_const_value(c: &Constant) -> (u32, u32) {
+    //     //....todo...only u32 for now... should also provide the sign
+    //     match c.value_type {
+    //         ObjectType::boolean => (if c.value.is_zero() { 0 } else { 1 }, 1),
+    //         ObjectType::native_field => todo!(), //to field(value), field:prime
+    //         ObjectType::signed(b) | ObjectType::unsigned(b) => {
+    //             (c.value.clone().try_into().unwrap(), b)
+    //         }
+    //         _ => todo!(),
+    //     }
+    // }
+
+    pub fn get_const_value2(c: FieldElement, ctype: ObjectType) -> (u32, u32) {
+        //....todo...only u32 for now... should also provide the sign
+        match ctype {
+            ObjectType::boolean => (if c.is_zero() { 0 } else { 1 }, 1),
+            ObjectType::native_field => (c.to_u128().try_into().unwrap(), 32), //TODO! //to field(value), field:prime
+            ObjectType::signed(b) | ObjectType::unsigned(b) => (c.to_u128().try_into().unwrap(), b),
+            _ => todo!(),
+        }
+    }
+
+    // pub fn to_const(value: FieldElement, obj_type: ObjectType) -> NodeObj {
+    //     let c = Constant {
+    //         id: crate::ssa::code_gen::IRGenerator::dummy_id(),
+    //         value: BigUint::from_bytes_be(&value.to_bytes()),
+    //         value_str: String::new(),
+    //         value_type: obj_type,
+    //     };
+    //     NodeObj::Const(c)
+    // }
+
+    pub fn node_evaluate(
+        n: &NodeEval,
+        is_zero: &mut bool,
+        const_value: &mut Option<u32>,
+        bit_size: &mut u32,
+    ) {
+        match n {
+            &NodeEval::Const(c, t) => {
+                *is_zero = c.is_zero();
+                let cv = Instruction::get_const_value2(c, t);
+                *const_value = Some(cv.0);
+                *bit_size = cv.1;
+            }
+            _ => {
+                *const_value = None;
+            }
+        }
+    }
+
+    //Evalaute the instruction value when its operands are constant
+    pub fn evaluate(&self, lhs: &NodeEval, rhs: &NodeEval) -> NodeEval {
+        let mut r_is_zero = false;
+        let mut l_is_zero = false;
+        let mut l_constant = None;
+        let mut r_constant = None;
+        let mut r_bsize = 0;
+        let mut l_bsize = 0;
+        //let mut l_sign = false; //TODO
+        Instruction::node_evaluate(lhs, &mut l_is_zero, &mut l_constant, &mut l_bsize);
```

Comment:
Instead of passing in mutable variables, lets have this assosciated method return them instead

Reply:
There are times where we pass in a mutable reference, and the reason to do this is to save on heap allocations for large buffers. For example: https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_string

Even with that read_to_string, we have this method which wraps around it: https://doc.rust-lang.org/std/fs/fn.read_to_string.html . (Not to say people use the latter over the former)

I'd also not use this way, if the function is called once on a buffer because then you are not saving much. In this particular case, I don't think we are saving anything by passing in a `&mut Integer` and `&mut bool`. 

Note that we also do not do this in the rest of the code unless its for ownership reasons, its never because we want to optimise heap allocations, which seems to be a more common approach in C++ for example. (Maybe there are some places where we do need to do it due to large buffer re-allocations)

Reply:
Thank you for your answer, I have done the change. 
I think another important point is the syntax shortcut when returning several arguments. If you were required to generate a specific return type structure as you would be in C++ you won't do this!

Reply:
Good point!